pkg/steps/run_test.go

```diff
@@ -40,135 +46,289 @@ func (*fakeStep) Objects() []ctrlruntimeclient.Object { return nil }
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceRoot)},
-	}
-	other := &fakeStep{
-		name:      "other",
-		shouldRun: true,
-		requires:  []api.StepLink{api.ExternalImageLink(api.ImageStreamTagReference{Namespace: "ns", Name: "base", Tag: "other"})},
-		creates:   []api.StepLink{api.InternalImageLink("other")},
-	}
-	src := &fakeStep{
-		name:      "src",
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceRoot)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceSource)},
-	}
-	bin := &fakeStep{
-		name:      "bin",
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceSource)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceBinaries)},
-	}
-	testBin := &fakeStep{
-		name:      "testBin",
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceSource)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceTestBinaries)},
-	}
-	rpm := &fakeStep{
-		name:      "rpm",
-		runErr:    errors.New("oopsie"),
-		shouldRun: true,
-		requires:  []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceBinaries)},
-		creates:   []api.StepLink{api.InternalImageLink(api.PipelineImageStreamTagReferenceRPMs)},
-	}
-	unrelated := &fakeStep{
-		name:      "unrelated",
-		shouldRun: false,
-		requires:  []api.StepLink{api.InternalImageLink("other"), api.InternalImageLink(api.PipelineImageStreamTagReferenceRPMs)},
-		creates:   []api.StepLink{api.InternalImageLink("unrelated")},
-	}
-	final := &fakeStep{
-		name:      "final",
-		shouldRun: false,
-		requires:  []api.StepLink{api.InternalImageLink("unrelated")},
-		creates:   []api.StepLink{api.InternalImageLink("final")},
-	}
+			var steps []api.Step
+			for _, step := range tc.steps {
+				if step.runErr != nil {
+					shouldFail++
+				}
+				if step.shouldRun {
+					shouldRun++
+				}
+				steps = append(steps, step)
+			}
 
-	suites, _, err := Run(context.Background(), api.BuildGraph([]api.Step{root, other, src, bin, testBin, rpm, unrelated, final}))
-	if err == nil {
-		t.Error("got no error but expected one")
-	}
-	if suites.Suites != nil && len(suites.Suites) != 1 || len(suites.Suites[0].TestCases) != 6 || suites.Suites[0].NumTests != 6 || suites.Suites[0].NumFailed != 1 {
-		t.Errorf("unexpected junit output: %#v", suites.Suites[0])
-	}
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			if tc.cancelled {
+				cancel()
+			}
+			suites, _, errs := Run(ctx, api.BuildGraph(steps))
+			if errs == nil && len(tc.errExpected) > 0 {
+				t.Error("got no error but expected one")
+			}
+
+			if errs != nil && len(tc.errExpected) > 0 {
+				var expectedErrorsString []string
```

Comment:
Why all of this complex error unwrapping logic instead of a `reflect.DeepEqual` ?

Reply:
It panics because of unexported fields. I made it this way to make it more readable if a test fails. I will try to check this out though.

Reply:
Maybe we can use `cmp.Diff` and its ability to forciby compare (or ignore) unexported fields? https://godoc.org/github.com/google/go-cmp/cmp#AllowUnexported

Reply:
Thanks @petr-muller! I will check this out

Reply:
Well, it seems that `AllowUnexported` works only for structs. In our case, we have slices of structs. I wasn't able to find any solution. I guess we can stick with this approach.

Reply:
It definitely works over slices: https://play.golang.org/p/5ze2Rw6gwBT